api/proto/teleport/join/v1/joinservice.proto

```diff
@@ -0,0 +1,158 @@
+// Copyright 2025 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package teleport.join.v1;
+
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/join/v1;joinv1";
+
+// ClientInit is the first message sent from the client during the join process, it
+// holds parameters common to all join methods.
+message ClientInit {
```

Comment:
Nit: I'd almost be tempted to introduce a sub-message for Agent-specific fields (e.g `dns_names`) and Bot-specific fields (e.g `expires`) to help make it clear what is required in each case. WDYT?

Reply:
I like it, done in the latest